contracts/vaults/TreasuryVault/RibbonAutocallVault.sol

```diff
@@ -0,0 +1,461 @@
+        )
+    {
+        uint256 _observationPeriodFreq = observationPeriodFreq;
+        uint256 _numTotalObservationPeriods = numTotalObservationPeriods;
+
+        uint256 currentObservation =
+            (_expiry - block.timestamp) / _observationPeriodFreq + 1;
+        address underlying = vaultParams.underlying;
+
+        uint256 autocallObservation = currentObservation;
+
+        // For every previous observation timestamp
+        for (uint256 i = currentObservation; i > 0; i--) {
+            // Gets observation timestamp of observation index
+            uint256 observationPeriodTimestamp =
+                _expiry -
+                    (_numTotalObservationPeriods - i) *
+                    _observationPeriodFreq;
+
+            uint256 observationPeriodPrice =
+                ORACLE.getExpiryPrice(underlying, observationPeriodTimestamp);
+
+            // Check if autocallable
+            if (
+                observationPeriodPrice >=
+                (initialSpotPrice * couponState.autocallBarrierPCT) /
+                    PCT_MULTIPLIER
+            ) {
+                autocallObservation = i;
+                autocallTimestamp = observationPeriodTimestamp;
+                break;
+            }
+        }
+
+        // For every observation timestamp before autocall
+        for (uint256 i = autocallObservation; i > 0; i--) {
+            // Gets observation timestamp of observation index
+            uint256 observationPeriodTimestamp =
+                _expiry -
+                    (_numTotalObservationPeriods - i) *
+                    _observationPeriodFreq;
+
+            uint256 observationPeriodPrice =
+                ORACLE.getExpiryPrice(underlying, observationPeriodTimestamp);
+
+            // Check if coupon barrier hit
+            if (
+                observationPeriodPrice >=
+                (initialSpotPrice * couponState.couponBarrierPCT) /
+                    PCT_MULTIPLIER
+            ) {
+                numCouponsEarned += 1;
+                // Get latest observation to hit coupon barrier
+                if (lastCouponBarrierBreachObservation == 0) {
+                    lastCouponBarrierBreachObservation = i;
+                }
+            }
+        }
+    }
+
+    /**
+     * @dev Verifies the coupon state is valid
+     * @param _couponType is the coupon type
+     * @param _autocallBarrierPCT is the autocall barrier pct
+     * @param _couponBarrierPCT is the coupon barrier pct
+     */
+    function _verifyCouponState(
+        CouponType _couponType,
+        uint256 _autocallBarrierPCT,
+        uint256 _couponBarrierPCT
+    ) internal pure {
+        require(_autocallBarrierPCT > PCT_MULTIPLIER, "!_autocallBarrierPCT");
+
+        if (_couponType == CouponType.FIXED) {
+            // Coupon Barrier = 0
+            require(_couponBarrierPCT == 0, "!FIXED");
+        } else if (
+            _couponType == CouponType.VANILLA || _couponType == CouponType.FIXED
```

Comment:
is ` || _couponType == CouponType.FIXED` supposed to be on this else if condition? 

Given that for the coupon type FIXED - there seems to be a conflict since the 3 conditions below can't be true simultaneously:
- `_couponBarrierPCT == 0`
- `_couponBarrierPCT == _autocallBarrierPCT`
- `_autocallBarrierPCT > PCT_MULTIPLIER`